simplified-app-shared/src/main/java/org/nypl/simplified/app/ReaderActivitySyncManager.kt

```diff
@@ -0,0 +1,173 @@
+    if (!annotationsManager.syncIsPossibleAndPermitted()) {
+      delayPageSync = false
+      LOG.debug("Account does not support sync or sync is disabled.")
+      return
+    }
+    if (feedEntry == null || !feedEntry.annotations.isSome) {
+      delayPageSync = false
+      LOG.error("No annotations uri or feed entry exists for this book. Abandoning sync attempt.")
+      return
+    }
+
+    val uriString = (feedEntry.annotations as Some<URI>).get().toString()
+
+    annotationsManager.requestReadingPositionOnServer(entryID, uriString, { serverLocation ->
+
+      delayPageSync = false
+
+      if (serverLocation == null) {
+        LOG.debug("No server location object returned.")
+        return@requestReadingPositionOnServer
+      }
+
+      val alert = createAlertForSyncLocation(serverLocation, context) { shouldMove ->
+        if (shouldMove) {
+          navigateToLocation(serverLocation)
+        }
+      }
+
+      // Pass through without presenting the Alert Dialog if:
+      // 1 - The server and the client have the same page marked
+      // 2 - There is no recent page saved on the server
+      //TODO make sure this conditional is working...
+      if (currentLocation.toString() != serverLocation.toString()) {
+        UIThread.runOnUIThread {
+          alert.show()
+        }
+      }
+    })
+  }
+
+  private fun createAlertForSyncLocation(location: ReaderBookLocation, context: Context,
+                                       completion: (shouldMove: Boolean) -> Unit): AlertDialog
+  {
+    val builder= AlertDialog.Builder(context)
+    builder.setTitle("Sync Reading Position")
+
+    //TODO add cases once i figure out null content cfi equivalent from
+    if (bookPackage != null) {
+      val chapterTitle = bookPackage.getSpineItem(location.idRef).title
+      builder.setMessage("Would you like to go to the latest page read? \n\nChapter:\n\"${chapterTitle}\"")
+    } else {
+      builder.setMessage("Would you like to go to the latest page read?")
+    }
+
+    builder.setPositiveButton("YES") { dialog, which ->
+      completion(true)
+      LOG.debug("User chose to jump to synced page: ${location.contentCFI}")
+    }
+
+    builder.setNegativeButton("NO") { dialog, which ->
+      completion(false)
+      LOG.debug("User declined jump to synced page.")
+    }
+
+    return builder.create()
+  }
+
+
+  /**
+   * Attempt to update current page to the server.
+   * @param location The page to be sent as the current "left off" page
+   * @param uri Annotation URI for the specific entry
+   */
+  fun updateServerReadingLocation(location: ReaderBookLocation)
+  {
+    //TODO note to self: delayPageSync does not seem to work on second attempt.
+
+    if (delayPageSync) {
+      LOG.debug("Post of last read position delayed. Initial sync still in progress.")
+      return
+    }
+
+    if (annotationsManager.syncIsPossibleAndPermitted()) {
+      queueTimer.cancel()
+      queueTimer = Timer()
+      queueTimer.schedule(3000) {
```

Comment:
This was a direct port of the current rate-limiting code to server requests. It will be updated to reflect a more robust implementation to match iOS.